File: Presentation/TemplateSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using InterviewMasterApp.Domain.Model;
using InterviewMasterApp.Presentation.Windows;

namespace InterviewMasterApp.Presentation
{
    /// <summary>
    /// Template selector dialog for loading interview templates.
    /// Ported from Swift: TemplateSelector.swift
    /// Provides category tabs and template cards with questions preview.
    /// </summary>
    public class TemplateSelector
    {
        private readonly RichTextBox _notesTextView;
        private readonly Action _switchToNotesTabCallback;
        private readonly Action _renderMarkdownCallback;

        public TemplateSelector(
            RichTextBox notesTextView,
            Action switchToNotesTabCallback,
            Action renderMarkdownCallback)
        {
            _notesTextView = notesTextView ?? throw new ArgumentNullException(nameof(notesTextView));
            _switchToNotesTabCallback = switchToNotesTabCallback ?? throw new ArgumentNullException(nameof(switchToNotesTabCallback));
            _renderMarkdownCallback = renderMarkdownCallback ?? throw new ArgumentNullException(nameof(renderMarkdownCallback));
        }

        /// <summary>
        /// Show template selector dialog.
        /// </summary>
        public void ShowTemplateSelector(Window owner)
        {
            var dialog = new Window
            {
                Title = "Interview Templates",
                Width = 480,
                Height = 520,
                WindowStartupLocation = WindowStartupLocation.CenterOwner,
                Owner = owner,
                WindowStyle = WindowStyle.ToolWindow,
                ResizeMode = ResizeMode.CanResize,
                Background = new SolidColorBrush(Color.FromArgb(230, 40, 40, 45))
            };

            var mainGrid = new Grid();
            mainGrid.RowDefinitions.Add(new RowDefinition { Height = GridLength.Auto }); // Tab bar
            mainGrid.RowDefinitions.Add(new RowDefinition { Height = new GridLength(1, GridUnitType.Star) }); // Content

            // Category tabs
            var tabControl = new TabControl
            {
                Margin = new Thickness(20, 10, 20, 10),
                Background = Brushes.Transparent,
                BorderThickness = new Thickness(0)
            };
            Grid.SetRow(tabControl, 0);

            var categories = new[]
            {
                InterviewTemplate.Category.Behavioral,
                InterviewTemplate.Category.SystemDesign,
                InterviewTemplate.Category.Coding,
                InterviewTemplate.Category.LanguageSpecific
            };

            // Scroll view container (shared across tabs)
            var scrollViewer = new ScrollViewer
            {
                VerticalScrollBarVisibility = ScrollBarVisibility.Auto,
                HorizontalScrollBarVisibility = ScrollBarVisibility.Disabled,
                Margin = new Thickness(0, 0, 0, 10)
            };
            Grid.SetRow(scrollViewer, 1);

            var listContainer = new StackPanel
            {
                Margin = new Thickness(15, 10, 15, 10)
            };
            scrollViewer.Content = listContainer;

            // Create tabs
            foreach (var category in categories)
            {
                var tab = new TabItem
                {
                    Header = GetCategoryDisplayName(category),
                    Background = new SolidColorBrush(Color.FromArgb(128, 60, 60, 67)),
                    Foreground = Brushes.White,
                    FontSize = 13,
                    Padding = new Thickness(12, 6, 12, 6)
                };
                tabControl.Items.Add(tab);
            }

            // Handle tab selection changes
            tabControl.SelectionChanged += (s, e) =>
            {
                if (tabControl.SelectedIndex >= 0 && tabControl.SelectedIndex < categories.Length)
                {
                    var selectedCategory = categories[tabControl.SelectedIndex];
                    PopulateTemplateList(listContainer, selectedCategory, dialog);
                }
            };

            mainGrid.Children.Add(tabControl);
            mainGrid.Children.Add(scrollViewer);

            dialog.Content = mainGrid;

            // Populate initial category
            PopulateTemplateList(listContainer, categories[0], dialog);

            dialog.ShowDialog();
        }

        /// <summary>
        /// Populate template list for a specific category.
        /// </summary>
        private void PopulateTemplateList(StackPanel container, InterviewTemplate.Category category, Window dialog)
        {
            container.Children.Clear();

            var templates = BuiltInTemplates.TemplatesForCategory(category);

            foreach (var template in templates)
            {
                var card = CreateTemplateCard(template, dialog);
                container.Children.Add(card);
            }
        }

        /// <summary>
        /// Create a template card UI element.
        /// </summary>
        private Border CreateTemplateCard(InterviewTemplate template, Window dialog)
        {
            var cardHeight = 60 + template.Questions.Count * 28;

            var card = new Border
            {
                Background = new SolidColorBrush(Color.FromArgb(15, 255, 255, 255)), // 6% white
                CornerRadius = new CornerRadius(10),
                Margin = new Thickness(0, 0, 0, 10),
                Padding = new Thickness(15, 10, 15, 10),
                MinHeight = cardHeight
            };

            var grid = new Grid();
            grid.RowDefinitions.Add(new RowDefinition { Height = GridLength.Auto }); // Header
            grid.RowDefinitions.Add(new RowDefinition { Height = GridLength.Auto }); // Description
            grid.RowDefinitions.Add(new RowDefinition { Height = new GridLength(1, GridUnitType.Star) }); // Questions

            // Header row
            var headerGrid = new Grid();
            headerGrid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
            headerGrid.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });
            Grid.SetRow(headerGrid, 0);

            // Title
            var title = new TextBlock
            {
                Text = template.Name,
                Foreground = Brushes.White,
                FontSize = 14,
                FontWeight = FontWeights.SemiBold,
                VerticalAlignment = VerticalAlignment.Center
            };
            Grid.SetColumn(title, 0);
            headerGrid.Children.Add(title);

            // Load button
            var loadButton = new Button
            {
                Content = "Load",
                Width = 80,
                Height = 28,
                FontSize = 12,
                FontWeight = FontWeights.Medium,
                Background = new SolidColorBrush(Color.FromArgb(128, 0, 122, 255)), // Blue
                Foreground = Brushes.White,
                BorderThickness = new Thickness(0),
                Cursor = System.Windows.Input.Cursors.Hand
            };
            loadButton.Click += (s, e) => LoadTemplate(template, dialog);
            Grid.SetColumn(loadButton, 1);
            headerGrid.Children.Add(loadButton);

            grid.Children.Add(headerGrid);

            // Description
            var description = new TextBlock
            {
                Text = template.Description,
                Foreground = new SolidColorBrush(Color.FromArgb(128, 255, 255, 255)), // 50% white
                FontSize = 11,
                Margin = new Thickness(0, 4, 0, 8)
            };
            Grid.SetRow(description, 1);
            grid.Children.Add(description);

            // Questions list
            var questionsPanel = new StackPanel
            {
                Margin = new Thickness(0, 4, 0, 0)
            };
            Grid.SetRow(questionsPanel, 2);

            foreach (var question in template.Questions)
            {
                var questionGrid = new Grid
                {
                    Margin = new Thickness(0, 0, 0, 4)
                };
                questionGrid.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto }); // Dot
                questionGrid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) }); // Text

                // Difficulty dot
                var dot = new Ellipse
                {
                    Width = 6,
                    Height = 6,
                    Fill = GetDifficultyBrush(question.QuestionDifficulty),
                    VerticalAlignment = VerticalAlignment.Center,
                    Margin = new Thickness(5, 0, 10, 0)
                };
                Grid.SetColumn(dot, 0);
                questionGrid.Children.Add(dot);

                // Question text
                var questionText = new TextBlock
                {
                    Text = question.Text,
                    Foreground = new SolidColorBrush(Color.FromArgb(179, 255, 255, 255)), // 70% white
                    FontSize = 11,
                    TextTrimming = TextTrimming.CharacterEllipsis,
                    VerticalAlignment = VerticalAlignment.Center
                };
                Grid.SetColumn(questionText, 1);
                questionGrid.Children.Add(questionText);

                questionsPanel.Children.Add(questionGrid);
            }

            grid.Children.Add(questionsPanel);
            card.Child = grid;

            return card;
        }

        /// <summary>
        /// Load template into notes editor.
        /// </summary>
        private void LoadTemplate(InterviewTemplate template, Window dialog)
        {
            // Build markdown outline
            var markdown = $"# {template.Name}\n\n";
            markdown += $"*{template.Description}*\n\n";

            for (int i = 0; i < template.Questions.Count; i++)
            {
                var q = template.Questions[i];
                markdown += $"## {i + 1}. {q.Text}\n\n";
                markdown += $"**Topic:** {q.Topic} | **Difficulty:** {q.DifficultyDisplayName}\n\n";

                if (q.Hints.Count > 0)
                {
                    markdown += "**Hints:**\n";
                    foreach (var hint in q.Hints)
                    {
                        markdown += $"- {hint}\n";
                    }
                    markdown += "\n";
                }

                markdown += "**Your notes:**\n\n\n";
            }

            // Load into notes tab
            SetNotesText(markdown);
            _switchToNotesTabCallback?.Invoke();
            _renderMarkdownCallback?.Invoke();

            // Close the dialog
            dialog.Close();

            StealthLogger.Shared.Log($"📋 Loaded template: {template.Name}");
        }

        /// <summary>
        /// Set text in notes RichTextBox.
        /// </summary>
        private void SetNotesText(string text)
        {
            var textRange = new System.Windows.Documents.TextRange(
                _notesTextView.Document.ContentStart,
                _notesTextView.Document.ContentEnd);
            textRange.Text = text;
        }

        /// <summary>
        /// Get difficulty color brush.
        /// </summary>
        private Brush GetDifficultyBrush(TemplateQuestion.Difficulty difficulty)
        {
            return difficulty switch
            {
                TemplateQuestion.Difficulty.Easy => new SolidColorBrush(Color.FromRgb(52, 199, 89)), // Green
                TemplateQuestion.Difficulty.Medium => new SolidColorBrush(Color.FromRgb(255, 149, 0)), // Orange
                TemplateQuestion.Difficulty.Hard => new SolidColorBrush(Color.FromRgb(255, 59, 48)), // Red
                _ => Brushes.Gray
            };
        }

        /// <summary>
        /// Get category display name.
        /// </summary>
        private string GetCategoryDisplayName(InterviewTemplate.Category category)
        {
            return category switch
            {
                InterviewTemplate.Category.Behavioral => "Behavioral",
                InterviewTemplate.Category.SystemDesign => "System Design",
                InterviewTemplate.Category.Coding => "Coding",
                InterviewTemplate.Category.LanguageSpecific => "Language-Specific",
                _ => category.ToString()
            };
        }
    }
}

File: Presentation/TemplateSelector.swift
import Cocoa

@available(macOS 14.0, *)
extension InterviewMasterDelegate {

    @objc func showTemplateSelector() {
        let panel = NSPanel(
            contentRect: NSRect(x: 0, y: 0, width: 480, height: 520),
            styleMask: [.titled, .closable, .resizable],
            backing: .buffered,
            defer: false
        )
        panel.title = "Interview Templates"
        panel.isFloatingPanel = true
        panel.level = .floating
        panel.center()

        let contentView = NSView(frame: panel.contentView!.bounds)
        contentView.autoresizingMask = [.width, .height]
        panel.contentView = contentView

        // Background
        let bg = NSVisualEffectView(frame: contentView.bounds)
        bg.autoresizingMask = [.width, .height]
        bg.material = .hudWindow
        bg.blendingMode = .behindWindow
        bg.state = .active
        contentView.addSubview(bg)

        // Category tabs
        let tabBar = NSSegmentedControl(frame: NSRect(x: 20, y: contentView.bounds.height - 50, width: contentView.bounds.width - 40, height: 30))
        tabBar.segmentCount = InterviewTemplate.Category.allCases.count
        for (i, cat) in InterviewTemplate.Category.allCases.enumerated() {
            tabBar.setLabel(cat.displayName, forSegment: i)
            tabBar.setWidth(0, forSegment: i) // Auto-size
        }
        tabBar.selectedSegment = 0
        tabBar.segmentStyle = .roundRect
        tabBar.autoresizingMask = [.width, .minYMargin]
        contentView.addSubview(tabBar)

        // Scroll view for template list
        let scrollView = NSScrollView(frame: NSRect(x: 0, y: 0, width: contentView.bounds.width, height: contentView.bounds.height - 60))
        scrollView.autoresizingMask = [.width, .height]
        scrollView.hasVerticalScroller = true
        scrollView.drawsBackground = false
        contentView.addSubview(scrollView)

        let listContainer = NSView(frame: NSRect(x: 0, y: 0, width: scrollView.bounds.width, height: 1000))
        scrollView.documentView = listContainer

        // Populate with initial category
        populateTemplateList(in: listContainer, scrollView: scrollView, category: .behavioral, panel: panel)

        // Handle tab changes via timer-based polling (simple approach without NSSegmentedControl target)
        tabBar.target = self
        tabBar.action = #selector(templateCategoryChanged(_:))
        tabBar.identifier = NSUserInterfaceItemIdentifier("templateTabBar")

        // Store references for the callback
        objc_setAssociatedObject(panel, "listContainer", listContainer, .OBJC_ASSOCIATION_RETAIN)
        objc_setAssociatedObject(panel, "scrollView", scrollView, .OBJC_ASSOCIATION_RETAIN)
        objc_setAssociatedObject(tabBar, "panel", panel, .OBJC_ASSOCIATION_RETAIN)

        panel.makeKeyAndOrderFront(nil)
    }

    @objc func templateCategoryChanged(_ sender: NSSegmentedControl) {
        let categories = InterviewTemplate.Category.allCases
        guard sender.selectedSegment >= 0, sender.selectedSegment < categories.count else { return }
        let category = categories[sender.selectedSegment]

        guard let panel = objc_getAssociatedObject(sender, "panel") as? NSPanel,
              let listContainer = objc_getAssociatedObject(panel, "listContainer") as? NSView,
              let scrollView = objc_getAssociatedObject(panel, "scrollView") as? NSScrollView else { return }

        populateTemplateList(in: listContainer, scrollView: scrollView, category: category, panel: panel)
    }

    private func populateTemplateList(in container: NSView, scrollView: NSScrollView, category: InterviewTemplate.Category, panel: NSPanel) {
        container.subviews.forEach { $0.removeFromSuperview() }

        let templates = BuiltInTemplates.templatesForCategory(category)
        let width = scrollView.bounds.width
        var y: CGFloat = 10

        for template in templates {
            // Template card
            let cardHeight: CGFloat = CGFloat(60 + template.questions.count * 28)
            let card = NSView(frame: NSRect(x: 15, y: y, width: width - 30, height: cardHeight))
            card.wantsLayer = true
            card.layer?.cornerRadius = 8
            card.layer?.backgroundColor = NSColor.white.withAlphaComponent(0.055).cgColor
            card.layer?.borderWidth = 1
            card.layer?.borderColor = NSColor.white.withAlphaComponent(0.10).cgColor

            // Title
            let title = NSTextField(labelWithString: template.name)
            title.frame = NSRect(x: 15, y: cardHeight - 28, width: card.frame.width - 120, height: 20)
            title.font = .systemFont(ofSize: 14, weight: .semibold)
            title.textColor = .white
            card.addSubview(title)

            // Description
            let desc = NSTextField(labelWithString: template.description)
            desc.frame = NSRect(x: 15, y: cardHeight - 48, width: card.frame.width - 120, height: 16)
            desc.font = .systemFont(ofSize: 11)
            desc.textColor = NSColor.white.withAlphaComponent(0.5)
            card.addSubview(desc)

            // Load button
            let loadButton = NSButton(frame: NSRect(x: card.frame.width - 95, y: cardHeight - 35, width: 80, height: 28))
            loadButton.title = "Load"
            loadButton.bezelStyle = .rounded
            loadButton.font = .systemFont(ofSize: 12, weight: .medium)
            loadButton.contentTintColor = template.category == .testAutomation ? .systemCyan : .white
            loadButton.target = self
            loadButton.action = #selector(loadTemplate(_:))
            loadButton.identifier = NSUserInterfaceItemIdentifier(template.id)
            objc_setAssociatedObject(loadButton, "panel", panel, .OBJC_ASSOCIATION_RETAIN)
            card.addSubview(loadButton)

            // Questions list
            var qy = cardHeight - 60
            for q in template.questions {
                let difficultyColor: NSColor = {
                    switch q.difficulty {
                    case .easy: return NSColor.systemGreen
                    case .medium: return NSColor.systemOrange
                    case .hard: return NSColor.systemRed
                    }
                }()

                let dot = NSView(frame: NSRect(x: 20, y: qy + 4, width: 6, height: 6))
                dot.wantsLayer = true
                dot.layer?.cornerRadius = 3
                dot.layer?.backgroundColor = difficultyColor.cgColor
                card.addSubview(dot)

                let qLabel = NSTextField(labelWithString: q.text)
                qLabel.frame = NSRect(x: 32, y: qy, width: card.frame.width - 50, height: 16)
                qLabel.font = .systemFont(ofSize: 11)
                qLabel.textColor = NSColor.white.withAlphaComponent(0.7)
                qLabel.lineBreakMode = .byTruncatingTail
                card.addSubview(qLabel)

                qy -= 28
            }

            container.addSubview(card)
            y += cardHeight + 10
        }

        container.frame.size.height = max(scrollView.bounds.height, y + 10)
        scrollView.contentView.scroll(to: NSPoint(x: 0, y: 0))
    }

    @objc func loadTemplate(_ sender: NSButton) {
        guard let templateId = sender.identifier?.rawValue else { return }
        guard let template = BuiltInTemplates.all.first(where: { $0.id == templateId }) else { return }

        // Build markdown outline
        var markdown = "# \(template.name)\n\n"
        markdown += "*\(template.description)*\n\n"

        for (i, q) in template.questions.enumerated() {
            markdown += "## \(i + 1). \(q.text)\n\n"
            markdown += "**Topic:** \(q.topic) | **Difficulty:** \(q.difficulty.rawValue.capitalized)\n\n"
            if !q.hints.isEmpty {
                markdown += "**Hints:**\n"
                for hint in q.hints {
                    markdown += "- \(hint)\n"
                }
                markdown += "\n"
            }
            markdown += "**Your notes:**\n\n\n"
        }

        // Load into notes tab
        textView.string = markdown
        switchToNotesTab()
        renderMarkdown()

        // Close the panel
        if let panel = objc_getAssociatedObject(sender, "panel") as? NSPanel {
            panel.close()
        }
    }
}
